docs/guides/1227-contributing.md

```diff
@@ -0,0 +1,174 @@
+git checkout main
+
+# Update your fork's main branch from upstream
+git pull upstream main
+
+# Checkout your feature branch
+git checkout my_feature_branch
+
+# Rebase your feature branch changes on top of the updated main branch
+git rebase main
+
+# Update your pull request with latest changes
+git push -f my_feature_branch
+```
+
+## Scope of pull requests
+
+We prefer small incremental changes that can be reviewed and merged quickly.
+It's OK if it takes multiple pull requests to close an issue.
+
+The idea is that each improvement should land in Dagger's main branch within a
+few hours.  The sooner we can get multiple people looking at and agreeing on a
+specific change, the quicker we will have it out in a release. The quicker we
+can get these small improvementes in a Dagger release, the quicker we can get
+feedback from our users and find out what doesn't work, or what we have missed.
+
+The added benefit is that this will force everyone to think about handling
+partially implemented features & non-breaking changes. Both are great
+approaches, and they work really well in the context of Dagger.
+
+["Small incremental changes ftw"](https://github.com/dagger/dagger/pull/1348#issuecomment-1009628531) -> Small pull requests that get merged within hours!
+
+## Commits
+
+### DCO
+
+Contributions to this project must be accompanied by a Developer Certificate of
+Origin (DCO).
+
+All commit messages must contain the Signed-off-by line with an email address
+that matches the commit author. When commiting, use the `--signoff` flag:
+
+```shell
+git commit -s
+```
+
+The Signed-off-by line must match the **author's real name**, otherwise the PR will be rejected.
+
+### Commit messages
+
+:::tip
+[How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/)
+:::
+
+Guidelines:
+
+- **Group Commits:** Each commit should represent a meaningful change (e.g. implement feature X, fix bug Y, ...).
+  - For instance, a PR should not look like _1) Add Feature X 2) Fix Typo 3) Changes to features X 5) Bugfix for feature X 6) Fix Linter 7)_ ...
+  - Instead, these commits should be squashed together into a single "Add Feature" commit.
+- Each commit should work on its own: it must compile, pass the linter and so on.
+  - This makes life much easier when using `git log`, `git blame`, `git bisect`, etc.
+  - For instance, when doing a `git blame` on a file to figure out why a change
+  was introduced, it's pretty meaningless to see a _Fix linter_ commit message.
+  "Add Feature X" is much more meaningful.
+- Use `git rebase -i main` to group commits together and rewrite their commit message.
+- To add changes to the previous commit, use `git commit --amend -s`. This will
+  change the last commit (amend) instead of creating a new commit.
+- Format: Use the imperative mood in the subject line: "If applied, this commit
+  will _your subject line here_"
+- Add the following prefixes to your commit message to help trigger automated processes[^1]:
```

Comment:
What about `feat` and `fix`?

Reply:
Yeah, right now these are focused on changelog exclusion, but the explanation is generic so I get your question. Let's improve that in another PR.